### DISCUSSION TITLE
Trouble with global interpreter

### DISCUSSION THREAD

**Discussion by robml** [pdm-project/pdm#981] [Q&A]
Recently set up `pdm`s global interpreter. Initially I had trouble because I couldn't access `pdm` with `sudo` needed to access some files. I got around this by prepending this `sudo -E env "PATH=$PATH"` (via an alias) whenever I wanted to call `pdm` with the global `-g` flag. While it has worked great to be able to now add packages like `pandas` via `sudo -E env "PATH=$PATH" pdm add -g pandas`, I still face trouble when using the python interpreter that is linked to the global project.

Namely, when I run `sudo -E env "PATH=$PATH" pdm run -g python` and within the python shell execute `>>> import pandas` I get a `ModuleNotFoundError`. I am confused now as to why the packages installed on the global project were not found by the python interpreter linked to the global project? Any ideas?

**Comment by frostming** [comment]
Or it is better to use the absolute pdm path: `/absolute/path/to/pdm add -g pandas`.

For a global interpreter, you don't need `pdm run`. Just use the global interpreter.

**Reply by robml** [reply]
I might use this `/absolute/path/to/pdm add -g pandas` structure sure, I am not entirely sure about the tradeoff but I assume it is more explicit.

As for the global interpreter what do you mean? Simply just running `python` and within the interpreter running `>>> import pandas`? With or without `pdm run` I get a `ModuleNotFoundError` btw.

**Reply by frostming** [reply]
Check the global interpreter path selected by PDM: `pdm info -g`. For example if the path is `/usr/bin/python`, then the packages should be installed to `/usr/lib/python3.8/site-packages` (with `3.8` being the major and minor version). And if you run `/usr/bin/python`, all packages should be ready there.

**Reply by robml** [reply]
So I did a few checks, and very strangely, all the site packages are correctly in `/usr/lib/python3.8/site-packages`, similarly the interpreter uses `/usr/bin/python3.8` which even if I run `python3` defaults to that same file. The global project was created in its default directory as well and it seems in order. However even when running that same interpreter I still get a `ModuleNotFoundError`.

**Reply by frostming** [reply]
Debian systems are reading `/usr/local/lib/python3.8/dist-packages` and `/usr/lib/python3/dist-packages`, but no `/usr/lib/python3.8/site-packages`. PDM didn't support the paths for now.

**Reply by frostming** [reply]
https://github.com/pdm-project/pdm/issues/885 will solve this problem

**Reply by robml** [reply]
Thanks for this, do you mean I should opt to create a separate global project directory different from the default (that is what I understood from #885)? Also out of curiosity why doesn't the default directory point to `site-packages` is that just not included in PDM support for now or is that intentionally by design? As for creating a new project folder, is it necessary to uninstall the packages and re-add them or can I just copy and paste the `.lock` and `.toml` files found in `~/.pdm/global-project` to the new project folder?

**Reply by frostming** [reply]
It is due to the insufficient patch of `python` on Debian systems: By default `/usr/lib/python3.8/site-packages` is not in `sys.path` but PDM installs packages into that very location as told by `sysconfig.get_paths()`, which should have been patched to return the correct location `/usr/local/lib/python3.8/dist-packages`. This issue should be fixed on Python3.10+.

See also https://github.com/deadsnakes/issues/issues/182

**Reply by robml** [reply]
Interesting. I am confused however what was your recommendation exactly when you referenced #885? To create another global project folder and move the `.lock` and `.toml` files there? Would I need to call `pdm run -gp` every time I wanted to use the global interpreter with that particular folder?

**Reply by robml** [reply]
@frostming hope I am not disturbing, when you have the time could you clarify the above point?

**Reply by frostming** [reply]
That issue has been fixed and released.

Now you just need to `pdm config global_project.user_site true`, then `pdm install -g` will install all packages into the user site `~/.local`

**Reply by robml** [reply]
Oh amazing. If I had previously installed packages with `pdm install -g` should I uninstall them first before updating and running the commands above?

**Reply by frostming** [reply]
Not necessary if you don't mind duplicate copies of the same package.
